Xamarin.Forms.Platform.Android/Renderers/EntryRenderer.cs

```diff
@@ -281,45 +294,123 @@ void UpdateReturnType()
 		{
 			if (Control == null || Element == null)
 				return;
-			
+
 			Control.ImeOptions = Element.ReturnType.ToAndroidImeAction();
 			_currentInputImeFlag = Control.ImeOptions;
 		}
 
 		void SelectionChanged(object sender, SelectionChangedEventArgs e)
 		{
-			var control = Control;
-			if (control == null || Element == null)
+			if (_nativeSelectionIsUpdating || Control == null || Element == null)
 				return;
 
-			var start = Element.CursorPosition;
+			int cursorPosition = Element.CursorPosition;
+			int selectionStart = Control.SelectionStart;
 
-			if (control.SelectionStart != start)
-				ElementController?.SetValueFromRenderer(Entry.CursorPositionProperty, control.SelectionStart);
+			if (!_cursorPositionChangePending)
+			{
+				var start = cursorPosition;
 
-			var selectionLength = control.SelectionEnd - control.SelectionStart;
-			if (selectionLength != Element.SelectionLength)
-				ElementController?.SetValueFromRenderer(Entry.SelectionLengthProperty, selectionLength);
-		}
+				if (selectionStart != start)
+					SetCursorPositionFromRenderer(selectionStart);
+			}
 
+			if (!_selectionLengthChangePending)
+			{
+				int elementSelectionLength = System.Math.Min(Control.Text.Length - cursorPosition, Element.SelectionLength);
+
+				var controlSelectionLength = Control.SelectionEnd - selectionStart;
+				if (controlSelectionLength != elementSelectionLength)
+					SetSelectionLengthFromRenderer(controlSelectionLength);
+			}
+		}
 
 		void UpdateCursorSelection()
 		{
-			var control = Control;
-			if (control == null || Element == null)
+			if (_nativeSelectionIsUpdating || Control == null || Element == null)
 				return;
 
-			if (Element.IsSet(Entry.CursorPositionProperty) || Element.IsSet(Entry.SelectionLengthProperty))
+			if (Control.RequestFocus())
 			{
-				var start = Element.CursorPosition;
-				var end = System.Math.Min(control.Length(), Element.CursorPosition + Element.SelectionLength);
+				try
+				{
+					int start = GetSelectionStart();
+					int end = GetSelectionEnd(start);
 
-				if (control.SelectionStart != start || control.SelectionEnd != end)
+					Control.SetSelection(start, end);
+				}
+				catch (System.Exception ex)
 				{
-					control.SetSelection(start, end);
-					control.RequestFocus();
+					Internals.Log.Warning("Entry", $"Failed to set Control.Selection from CursorPosition/SelectionLength: {ex}");
 				}
+				finally
+				{
+					_cursorPositionChangePending = _selectionLengthChangePending = false;
+				}
+			}
+		}
+
+		int GetSelectionEnd(int start)
+		{
+			int end = start;
+			int selectionLength = Element.SelectionLength;
+
+			if (Element.IsSet(Entry.SelectionLengthProperty))
```

Comment:
if its not set can't we just assume the value is the default value and not do the lookup?

Reply:
What would the default value be when the value was set but is now cleared?

Reply:
SelectionLengthProperty.DefaultValue?